[JUJU-980] Add FSActionsChecker and FSAccessRule to safeio

arnodel opened the pull request:
This allows implementing access checking in safeio file access functions
An example of a simple prefix rule is implemented (PrefixFSAccessRule)
Also a rule to compose rules (FSAccessRuleset)

runtime.RuntimeContextDef can now take a FSAccessRule to specify what
access is allowed to the filesystem.

arnodel committed:
Add FSActionsChecker and FSAccessRule to safeio

This allows implementing access checking in safeio file access functions
An example of a simple prefix rule is implemented (PrefixFSAccessRule)
Also a rule to compose rules (FSAccessRuleset)

runtime.RuntimeContextDef can now take a FSAccessRule to specify what
access is allowed to the filesystem.

arnodel committed:
Lua tests for fs access rules

arnodel committed:
More tests for fs access

arnodel committed:
A bit of documentation

arnodel committed:
Chain fs access rules in context children

This is so that permissions cannot increase

arnodel committed:
Fix noquotas

arnodel committed:
Try explicitly closing files to make tests pass on windows

cderici commented on safeio/rules.go:
```diff
@@ -0,0 +1,100 @@
+func (r PrefixFSAccessRule) GetFSAccessEffect(path string, actions FSAction) (allowed FSAction, denied FSAction) {
+	if !strings.HasPrefix(path, r.Prefix) {
+		// If the path does start with r.Prefix, there is no effect from this rule.
+		return 0, 0
```
is returning 0 for `allowed` when the path doesn't start with the prefix going to be a problem for the `FSAccessRulechain`, since it `&=` s the rules? (I see it might false the whole thing where it shouldn't have any effect, right?, or am I missing something here?)

Also you probably meant `... doesn't ...` in the comment that says `... does start with r.Prefix`

arnodel commented:
I think this is the desired behaviour in the context of the rule chain. The
rule does not explicitly show access so the chain it belongs to should not
either.

About the content: yes you are right, will fix that.

On Mon, 18 Apr 2022, 23:04 Caner Derici, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In safeio/rules.go
> <https://github.com/canonical/golua/pull/6#discussion_r852445124>:
>
> > +		allowed &= a
> +		denied |= d
> +	}
> +	return
> +}
> +
> +type PrefixFSAccessRule struct {
> +	Prefix         string
> +	AllowedActions FSAction
> +	DeniedActions  FSAction
> +}
> +
> +func (r PrefixFSAccessRule) GetFSAccessEffect(path string, actions FSAction) (allowed FSAction, denied FSAction) {
> +	if !strings.HasPrefix(path, r.Prefix) {
> +		// If the path does start with r.Prefix, there is no effect from this rule.
> +		return 0, 0
>
> is returning 0 for allowed when the path doesn't start with the prefix
> going to be a problem for the FSAccessRulechain, since it &= s the rules?
> (I see it might false the whole thing where it shouldn't have any effect,
> right?, or am I missing something here?)
>
> Also you probably meant ... doesn't ... in the comment that says ... does
> start with r.Prefix
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical/golua/pull/6#pullrequestreview-944867722>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMJKOO66IUFLCV5VPHTQT3VFXMAFANCNFSM5TN43JOA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

arnodel committed:
Add tags for test files

This allows having test files running only on some platforms

arnodel committed:
Try to make tags pattern work on windows

arnodel committed:
Ensure gc operates per runtime (#84)

* Ensure gc operates per runtime
* Ensure gc operates per runtime

* Make sure UserData has its __gc metamethod registered

* Document how Golua runtime uses weak refs.

* Implement isolating gc in resource constrained contexts

* Runtime.Close tries to clean all contexts

* Add sections in quotaas doc about finalizers

* Separate quotas and non quotas tests

* Tests for UnsafePool

* Tests for SafePool

* ClonePool implementation

This implementation of luagc.Pool  allows GC to happen any time
but does not support weakrefs

* Add convenience functions for running code with quotas

arnodel committed:
Fix embedding quickstart example - fixes #92 (#94)

sammy-ette committed:
Implement io.popen (#88)

arnodel committed:
Allow os.date with no argument (#95)

Fixes #91

arnodel committed:
Relax time-sensitive test

TheCount committed:
Support equality comparison of userdata (#96)

Co-authored-by: Alexander Klauer <Alexander.Klauer@googlemail.com>

TheCount committed:
Fix overflow on 32-bit archs (#98)

* Fix overflow on 32-bit archs

packFormatReader uses an uint for its option size, so maxDecuplable must
not overflow uint.

* Properly format integer value as string

On architectures where int is narrower than int64, integer values
would be formatted incorrectly when converted to string with
strconv.Itoa. Using strconv.FormatInt fixes this.

TheCount committed:
Forbid newindex for non-tables without __newindex (#99)

Previously, a newindex event on a non-table without a suitable
__newindex entry in its metatable would be silently ignored. This commit
causes an error to be raised.

arnodel committed:
Fix use of disallowed internal std library functions in go 1.23 (#106)

* Fix use of disallowed internal std library functions in go 1.23

* Update workflow

* update go version to test to 1.23

* Fix flaky tests

arnodel committed:
Increase size of continuation PC (#109)

Prevents a panic on executing large chunks

arnodel committed:
Update documentation of Table.Next (#107)

arnodel committed:
Introduce FSAccessEffect for clearer code
